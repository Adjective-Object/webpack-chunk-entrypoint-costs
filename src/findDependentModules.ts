import { ModuleGraphWithReasons } from 'webpack-bundle-diff-add-reasons';

export function findDependentModules(
    bundleGraph: ModuleGraphWithReasons,
    entrypoint: string
): string[] {
    let traversedModules = new Set<string>();
    let frontier = [entrypoint];
    while (frontier.length) {
        const thisModuleName = frontier.pop();
        if (!thisModuleName) {
            throw new Error('non-emtpy module frontier popped undefined');
        }
        traversedModules.add(thisModuleName);

        const thisNode = bundleGraph[thisModuleName];
        if (!thisNode) {
            throw new Error(
                `could not get module node for node name ${thisNode}`
            );
        }

        const nextChildren = thisNode.reasonChildren.filter(
            nodeName =>
                frontier.indexOf(nodeName) === -1 &&
                !traversedModules.has(nodeName)
        );
        frontier = frontier.concat(nextChildren);
    }

    return Array.from(traversedModules);
}
