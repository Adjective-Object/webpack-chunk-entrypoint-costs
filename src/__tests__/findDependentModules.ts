import { findDependentModules } from '../findDependentModules';
import { join as joinPath } from 'path';
import { readFile } from 'mz/fs';
import { ModuleGraphWithReasons } from 'webpack-bundle-diff-add-reasons';

describe('findDependentModules', () => {
    // let inputBundleData: { code: string | Buffer; map: string | Buffer };
    let bundleGraph: ModuleGraphWithReasons;

    beforeEach(async () => {
        bundleGraph = JSON.parse(
            await readFile(
                joinPath(__dirname, 'test-data', 'derived', 'bundleStats.json'),
                'utf-8',
            ),
        ).graph;
    });

    it('loads the bundlestats correctly', () => {
        expect(bundleGraph).not.toBe(undefined);
    });

    it('finds dependent modules', () => {
        const dependencies = findDependentModules(
            bundleGraph,
            './sample/script/ribbonButtonRenderer.tsx',
        );
        expect(dependencies).toEqual(
            jasmine.arrayContaining(['./node_modules/react/index.js']),
        );
    });

    it('does not find parent modules', () => {
        const dependencies = findDependentModules(
            bundleGraph,
            './sample/script/ribbonButtonRenderer.tsx',
        );
        expect(dependencies).not.toEqual(
            jasmine.arrayContaining(['./sample/script/sample.tsx']),
        );
    });

    it('matches the snapshot', () => {
        const result = findDependentModules(
            bundleGraph,
            './sample/script/ribbonButtonRenderer.tsx',
        );
        expect(result).toMatchSnapshot();
    });
});
