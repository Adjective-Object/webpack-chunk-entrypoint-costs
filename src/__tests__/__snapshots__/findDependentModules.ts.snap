// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findDependentModules matches the snapshot 1`] = `
Array [
  "./sample/script/ribbonButtonRenderer.tsx",
  "./sample/script/icons/unlink.svg",
  "./sample/script/icons/undo.svg",
  "./sample/script/icons/underline.svg",
  "./sample/script/icons/textcolor.svg",
  "./sample/script/icons/superscript.svg",
  "./sample/script/icons/subscript.svg",
  "./sample/script/icons/strikethrough.svg",
  "./sample/script/icons/rtl.svg",
  "./sample/script/icons/removeformat.svg",
  "./sample/script/icons/redo.svg",
  "./sample/script/icons/outdent.svg",
  "./sample/script/icons/outdent-rtl.svg",
  "./sample/script/icons/numbering.svg",
  "./sample/script/icons/numbering-rtl.svg",
  "./sample/script/icons/ltr.svg",
  "./sample/script/icons/italic.svg",
  "./sample/script/icons/indent.svg",
  "./sample/script/icons/indent-rtl.svg",
  "./sample/script/icons/imagealttext.svg",
  "./sample/script/icons/fontsize.svg",
  "./sample/script/icons/fontname.svg",
  "./sample/script/icons/emoji.svg",
  "./sample/script/icons/dropdown.svg",
  "./sample/script/icons/createlink.svg",
  "./sample/script/icons/bullets.svg",
  "./sample/script/icons/bullets-rtl.svg",
  "./sample/script/icons/bold.svg",
  "./sample/script/icons/blockquote.svg",
  "./sample/script/icons/backcolor.svg",
  "./sample/script/icons/alignright.svg",
  "./sample/script/icons/alignleft.svg",
  "./sample/script/icons/aligncenter.svg",
  "./node_modules/react/index.js",
  "./node_modules/react/cjs/react.production.min.js",
  "./node_modules/object-assign/index.js",
  "./node_modules/office-ui-fabric-react/lib/Image.js",
  "./node_modules/office-ui-fabric-react/lib/components/Image/index.js",
  "./node_modules/office-ui-fabric-react/lib/components/Image/Image.types.js",
  "./node_modules/office-ui-fabric-react/lib/components/Image/Image.js",
  "./node_modules/office-ui-fabric-react/lib/components/Image/Image.styles.js",
  "./node_modules/office-ui-fabric-react/lib/Styling.js",
  "./node_modules/office-ui-fabric-react/lib/version.js",
  "./node_modules/@uifabric/utilities/lib/dom/getWindow.js",
  "./node_modules/@uifabric/utilities/lib/dom/setSSR.js",
  "./node_modules/@uifabric/styling/lib/index.js",
  "./node_modules/@uifabric/styling/lib/version.js",
  "./node_modules/@uifabric/set-version/lib/index.js",
  "./node_modules/@uifabric/set-version/lib/setVersion.js",
  "./node_modules/@uifabric/styling/lib/utilities/index.js",
  "./node_modules/@uifabric/styling/lib/utilities/icons.js",
  "./node_modules/@uifabric/utilities/lib/warn/warn.js",
  "./node_modules/@uifabric/utilities/lib/GlobalSettings.js",
  "./node_modules/@uifabric/merge-styles/lib/mergeStyles.js",
  "./node_modules/@uifabric/merge-styles/lib/styleToClassName.js",
  "./node_modules/@uifabric/merge-styles/lib/transforms/rtlifyRules.js",
  "./node_modules/@uifabric/merge-styles/lib/transforms/provideUnits.js",
  "./node_modules/@uifabric/merge-styles/lib/transforms/prefixRules.js",
  "./node_modules/@uifabric/merge-styles/lib/getVendorSettings.js",
  "./node_modules/@uifabric/merge-styles/lib/transforms/kebabRules.js",
  "./node_modules/@uifabric/merge-styles/lib/extractStyleParts.js",
  "./node_modules/@uifabric/merge-styles/lib/StyleOptionsState.js",
  "./node_modules/@uifabric/merge-styles/lib/fontFace.js",
  "./node_modules/@uifabric/merge-styles/lib/Stylesheet.js",
  "./node_modules/tslib/tslib.es6.js",
  "./node_modules/@uifabric/styling/lib/utilities/getIconClassName.js",
  "./node_modules/@uifabric/styling/lib/utilities/buildClassMap.js",
  "./node_modules/@uifabric/styling/lib/styles/index.js",
  "./node_modules/@uifabric/styling/lib/styles/zIndexes.js",
  "./node_modules/@uifabric/styling/lib/styles/theme.js",
  "./node_modules/@uifabric/utilities/lib/merge.js",
  "./node_modules/@uifabric/utilities/lib/customizations/Customizations.js",
  "./node_modules/@uifabric/styling/lib/styles/DefaultSpacing.js",
  "./node_modules/@uifabric/styling/lib/styles/scheme.js",
  "./node_modules/@uifabric/utilities/lib/customizations/mergeSettings.js",
  "./node_modules/@uifabric/styling/lib/styles/hiddenContentStyle.js",
  "./node_modules/@uifabric/styling/lib/styles/getPlaceholderStyles.js",
  "./node_modules/@uifabric/styling/lib/styles/getGlobalClassNames.js",
  "./node_modules/@uifabric/utilities/lib/memoize.js",
  "./node_modules/@uifabric/merge-styles/lib/index.js",
  "./node_modules/@uifabric/merge-styles/lib/version.js",
  "./node_modules/@uifabric/merge-styles/lib/mergeStyleSets.js",
  "./node_modules/@uifabric/merge-styles/lib/keyframes.js",
  "./node_modules/@uifabric/merge-styles/lib/concatStyleSetsWithProps.js",
  "./node_modules/@uifabric/merge-styles/lib/concatStyleSets.js",
  "./node_modules/@uifabric/styling/lib/styles/getFocusStyle.js",
  "./node_modules/@uifabric/utilities/lib/initializeFocusRects.js",
  "./node_modules/@uifabric/utilities/lib/keyboard.js",
  "./node_modules/@uifabric/utilities/lib/KeyCodes.js",
  "./node_modules/@uifabric/styling/lib/styles/getFadedOverflowStyle.js",
  "./node_modules/@uifabric/styling/lib/styles/fonts.js",
  "./node_modules/@uifabric/styling/lib/styles/PulsingBeaconAnimationStyles.js",
  "./node_modules/@uifabric/styling/lib/styles/GeneralStyles.js",
  "./node_modules/@uifabric/styling/lib/styles/DefaultPalette.js",
  "./node_modules/@uifabric/styling/lib/styles/DefaultFontStyles.js",
  "./node_modules/@uifabric/utilities/lib/language.js",
  "./node_modules/@uifabric/utilities/lib/localStorage.js",
  "./node_modules/@uifabric/styling/lib/styles/DefaultEffects.js",
  "./node_modules/@uifabric/styling/lib/styles/CommonStyles.js",
  "./node_modules/@uifabric/styling/lib/styles/AnimationStyles.js",
  "./node_modules/@uifabric/styling/lib/classNames/index.js",
  "./node_modules/@uifabric/styling/lib/classNames/FontClassNames.js",
  "./node_modules/@uifabric/styling/lib/classNames/ColorClassNames.js",
  "./node_modules/@uifabric/styling/lib/classNames/AnimationClassNames.js",
  "./node_modules/@uifabric/styling/lib/MergeStyles.js",
  "./node_modules/office-ui-fabric-react/lib/Utilities.js",
  "./node_modules/@uifabric/utilities/lib/index.js",
  "./node_modules/@uifabric/utilities/lib/warn.js",
  "./node_modules/@uifabric/utilities/lib/warn/warnControlledUsage.js",
  "./node_modules/@uifabric/utilities/lib/version.js",
  "./node_modules/@uifabric/utilities/lib/string.js",
  "./node_modules/@uifabric/utilities/lib/setFocusVisibility.js",
  "./node_modules/@uifabric/utilities/lib/selection/index.js",
  "./node_modules/@uifabric/utilities/lib/selection/Selection.types.js",
  "./node_modules/@uifabric/utilities/lib/selection/Selection.js",
  "./node_modules/@uifabric/utilities/lib/scroll.js",
  "./node_modules/@uifabric/utilities/lib/dom/getDocument.js",
  "./node_modules/@uifabric/utilities/lib/safeSetTimeout.js",
  "./node_modules/@uifabric/utilities/lib/safeRequestAnimationFrame.js",
  "./node_modules/@uifabric/utilities/lib/rtl.js",
  "./node_modules/@uifabric/utilities/lib/sessionStorage.js",
  "./node_modules/@uifabric/utilities/lib/resources.js",
  "./node_modules/@uifabric/utilities/lib/properties.js",
  "./node_modules/@uifabric/utilities/lib/overflow.js",
  "./node_modules/@uifabric/utilities/lib/osDetector.js",
  "./node_modules/@uifabric/utilities/lib/object.js",
  "./node_modules/@uifabric/utilities/lib/mobileDetector.js",
  "./node_modules/@uifabric/utilities/lib/math.js",
  "./node_modules/@uifabric/utilities/lib/initials.js",
  "./node_modules/@uifabric/utilities/lib/initializeComponentRef.js",
  "./node_modules/@uifabric/utilities/lib/ie11Detector.js",
  "./node_modules/@uifabric/utilities/lib/hoistStatics.js",
  "./node_modules/@uifabric/utilities/lib/hoist.js",
  "./node_modules/@uifabric/utilities/lib/getId.js",
  "./node_modules/@uifabric/utilities/lib/focus.js",
  "./node_modules/@uifabric/utilities/lib/dom/getParent.js",
  "./node_modules/@uifabric/utilities/lib/dom/getVirtualParent.js",
  "./node_modules/@uifabric/utilities/lib/dom/isVirtualElement.js",
  "./node_modules/@uifabric/utilities/lib/dom/elementContainsAttribute.js",
  "./node_modules/@uifabric/utilities/lib/dom/findElementRecursive.js",
  "./node_modules/@uifabric/utilities/lib/dom/elementContains.js",
  "./node_modules/@uifabric/utilities/lib/extendComponent.js",
  "./node_modules/@uifabric/utilities/lib/dom.js",
  "./node_modules/@uifabric/utilities/lib/dom/setVirtualParent.js",
  "./node_modules/@uifabric/utilities/lib/dom/setPortalAttribute.js",
  "./node_modules/@uifabric/utilities/lib/dom/raiseClick.js",
  "./node_modules/@uifabric/utilities/lib/dom/portalContainsElement.js",
  "./node_modules/@uifabric/utilities/lib/dom/on.js",
  "./node_modules/@uifabric/utilities/lib/dom/getRect.js",
  "./node_modules/@uifabric/utilities/lib/dom/getChildren.js",
  "./node_modules/@uifabric/utilities/lib/customizations/mergeCustomizations.js",
  "./node_modules/@uifabric/utilities/lib/customizations/customizable.js",
  "./node_modules/@uifabric/utilities/lib/customizations/CustomizerContext.js",
  "./node_modules/@uifabric/utilities/lib/customizations/Customizer.js",
  "./node_modules/@uifabric/utilities/lib/css.js",
  "./node_modules/@uifabric/utilities/lib/controlled.js",
  "./node_modules/@uifabric/utilities/lib/classNamesFunction.js",
  "./node_modules/@uifabric/utilities/lib/assertNever.js",
  "./node_modules/@uifabric/utilities/lib/asAsync.js",
  "./node_modules/@uifabric/utilities/lib/array.js",
  "./node_modules/@uifabric/utilities/lib/aria.js",
  "./node_modules/@uifabric/utilities/lib/appendFunction.js",
  "./node_modules/@uifabric/utilities/lib/Rectangle.js",
  "./node_modules/@uifabric/utilities/lib/FabricPerformance.js",
  "./node_modules/@uifabric/utilities/lib/EventGroup.js",
  "./node_modules/@uifabric/utilities/lib/DelayedRender.js",
  "./node_modules/@uifabric/utilities/lib/BaseComponent.js",
  "./node_modules/@uifabric/utilities/lib/warn/warnMutuallyExclusive.js",
  "./node_modules/@uifabric/utilities/lib/warn/warnDeprecations.js",
  "./node_modules/@uifabric/utilities/lib/warn/warnConditionallyRequiredProps.js",
  "./node_modules/@uifabric/utilities/lib/initializeDir.js",
  "./node_modules/@uifabric/utilities/lib/AutoScroll.js",
  "./node_modules/@uifabric/utilities/lib/Async.js",
  "./node_modules/@uifabric/utilities/lib/styled.js",
  "./node_modules/office-ui-fabric-react/lib/components/Image/Image.base.js",
]
`;
